tests/unit/test_int8.cpp

```diff
@@ -0,0 +1,995 @@
+}
+
+TEST_F(INT8HNSWTest, searchByID) {
+    HNSWParams params = {.dim = 4, .M = 16, .efConstruction = 200};
+    EXPECT_NO_FATAL_FAILURE(search_by_id_test(params));
+}
+
+TEST_F(INT8BruteForceTest, searchByID) {
+    BFParams params = {.dim = 4};
+    EXPECT_NO_FATAL_FAILURE(search_by_id_test(params));
+}
+
+TEST_F(INT8TieredTest, searchByID) {
+    HNSWParams params = {.dim = 4, .M = 16, .efConstruction = 200};
+    EXPECT_NO_FATAL_FAILURE(search_by_id_test(params));
+}
+
+template <typename params_t>
+void INT8Test::search_by_score_test(params_t index_params) {
+    SetUp(index_params);
+
+    size_t k = 11;
+    size_t n = 100;
+
+    for (size_t i = 0; i < n; i++) {
+        this->GenerateAndAddVector(i, i); // {i, i, i, i}
+    }
+    ASSERT_EQ(VecSimIndex_IndexSize(index), n);
+
+    int8_t query[dim];
+    this->GenerateVector(query, 50); // {50, 50, 50, 50}
+
+    // Vectors values are equal to the id, so the 11 closest vectors are
+    // 45, 46...50 (closest), 51...55
+    static size_t expected_res_order[] = {50, 49, 51, 48, 52, 47, 53, 46, 54, 45, 55};
+    auto verify_res = [&](size_t id, double score, size_t index) {
+        ASSERT_EQ(id, expected_res_order[index]);
+        ASSERT_EQ(score, 4 * (50 - id) * (50 - id)); // L2 distance
+    };
+
+    // Search by score
+    runTopKSearchTest(index, query, k, verify_res);
+}
+
+TEST_F(INT8HNSWTest, searchByScore) {
+    HNSWParams params = {.dim = 4, .M = 16, .efConstruction = 200};
+    EXPECT_NO_FATAL_FAILURE(search_by_score_test(params));
+}
+
+TEST_F(INT8BruteForceTest, searchByScore) {
+    BFParams params = {.dim = 4};
+    EXPECT_NO_FATAL_FAILURE(search_by_score_test(params));
+}
+
+TEST_F(INT8TieredTest, searchByScore) {
+    HNSWParams params = {.dim = 4, .M = 16, .efConstruction = 200};
+    EXPECT_NO_FATAL_FAILURE(search_by_score_test(params));
+}
+
+template <typename params_t>
+void INT8Test::metrics_test(params_t index_params) {
+    SetUp(index_params);
+    size_t n = 10;
+    VecSimMetric metric = index_params.metric;
+    double expected_score = 0;
+
+    auto verify_res = [&](size_t id, double score, size_t index) {
+        ASSERT_EQ(score, expected_score) << "failed at vector id:" << id;
+    };
+
+    for (size_t i = 0; i < n; i++) {
+        int8_t vector[dim];
+        this->PopulateRandomAndAddVector(i, vector);
+
+        if (metric == VecSimMetric_Cosine) {
+            // compare with the norm stored in the index vector
+            const int8_t *index_vector = static_cast<const int8_t *>(this->GetDataByInternalId(i));
+            float index_vector_norm = *(reinterpret_cast<const float *>(index_vector + dim));
+            float vector_norm = spaces::IntegralType_ComputeNorm<int8_t>(vector, dim);
+            ASSERT_EQ(index_vector_norm, vector_norm) << "wrong vector norm for vector id:" << i;
+        } else if (metric == VecSimMetric_IP) {
+            expected_score = INT8_InnerProduct(vector, vector, dim);
```

Comment:
Why do we compute `expected_score` here and only for IP?

Reply:
because IP returns IP(v,v) and not the distance if the vector is not normalized, i.e dist(v,v) != 0

Reply:
Right... BTW, aren't we returning 1-IP(v,v)?

Reply:
we are, my bad